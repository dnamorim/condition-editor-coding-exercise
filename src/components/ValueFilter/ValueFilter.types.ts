import { Property } from 'data/datastore.types';

export type PropTypes = {
  placeholder: string;
  operator?: string;
  property?: Property;
  value?: string;
  onChange: (e: any, opt: any) => void;
};
