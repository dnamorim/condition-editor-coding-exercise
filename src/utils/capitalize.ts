const capitalize = (text: string | undefined) => {
  if (!text) {
    return text;
  }

  return text[0].toUpperCase() + text.slice(1).toLowerCase();
};

export default capitalize;
