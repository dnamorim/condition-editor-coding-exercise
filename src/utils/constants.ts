/* Property Types */
export const PROPERTY_TYPES = {
  STRING: 'string',
  NUMBER: 'number',
  ENUM: 'enumerated',
};

/* Operator Types */
export const OPERATOR_TYPES = {
  EQUALS: 'equals',
  GREATER_THAN: 'greater_than',
  LESS_THAN: 'less_than',
  ANY: 'any',
  NONE: 'none',
  IN: 'in',
  CONTAINS: 'contains',
};
