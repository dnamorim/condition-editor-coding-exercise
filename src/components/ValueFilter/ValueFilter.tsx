import { Dropdown, Input } from 'semantic-ui-react';
import { OPERATOR_TYPES, PROPERTY_TYPES } from 'utils/constants';
import { PropTypes } from './ValueFilter.types';

const ValueFilter = ({
  placeholder,
  operator,
  property,
  onChange,
  value,
  ...otherProps
}: PropTypes) => {
  if (
    !property ||
    !operator ||
    operator === '' ||
    operator === OPERATOR_TYPES.ANY ||
    operator === OPERATOR_TYPES.NONE
  ) {
    return <></>;
  }

  if (property.type === PROPERTY_TYPES.ENUM) {
    const options =
      property?.values?.map((value: string) => ({
        key: value,
        text: value,
        value,
      })) ?? [];

    return (
      <Dropdown
        fluid
        selection
        placeholder={placeholder}
        multiple={operator === OPERATOR_TYPES.IN}
        options={options}
        onChange={onChange}
        {...otherProps}
      />
    );
  }

  return (
    <Input
      fluid
      placeholder={placeholder}
      onChange={onChange}
      value={value}
      {...otherProps}
    />
  );
};

export default ValueFilter;
