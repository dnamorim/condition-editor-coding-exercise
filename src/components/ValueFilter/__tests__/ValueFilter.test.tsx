import { render, screen } from '@testing-library/react';
import ValueFilter from '..';
import { OPERATOR_TYPES, PROPERTY_TYPES } from 'utils/constants';

describe('ValueFilter', () => {
  const mockProps = {
    placeholder: 'Value(s)',
    onChange: jest.fn(),
    property: {
      id: 1,
      name: 'Product Name',
      type: PROPERTY_TYPES.STRING,
    },
  };

  describe('should not render', () => {
    it('when no property and operation is provided', () => {
      const { property, ...props } = mockProps;
      const { container } = render(<ValueFilter {...props} />);

      expect(container).toBeEmptyDOMElement();
    });

    it('when no operator is provided', () => {
      const { container } = render(<ValueFilter {...mockProps} />);

      expect(container).toBeEmptyDOMElement();
    });

    it('when operator is an empty string', () => {
      const operator = '';

      const { container } = render(
        <ValueFilter operator={operator} {...mockProps} />,
      );

      expect(container).toBeEmptyDOMElement();
    });

    it(`when operator is "${OPERATOR_TYPES.ANY}"`, () => {
      const { container } = render(
        <ValueFilter operator={OPERATOR_TYPES.ANY} {...mockProps} />,
      );

      expect(container).toBeEmptyDOMElement();
    });

    it(`when operator is "${OPERATOR_TYPES.NONE}"`, () => {
      const { container } = render(
        <ValueFilter operator={OPERATOR_TYPES.NONE} {...mockProps} />,
      );

      expect(container).toBeEmptyDOMElement();
    });
  });

  it(`should render a Dropdown when property type is ${PROPERTY_TYPES.ENUM}`, () => {
    const enumProperty = {
      id: 1,
      name: 'Product Type',
      type: PROPERTY_TYPES.ENUM,
      values: ['foo', 'bar'],
    };
    const operator = OPERATOR_TYPES.EQUALS;
    const props = {
      ...mockProps,
      property: enumProperty,
      operator,
    };
    const { container } = render(<ValueFilter {...props} />);

    expect(screen.getByRole('listbox')).toBeInTheDocument();
    expect(screen.queryAllByRole('option')).toHaveLength(
      enumProperty.values.length,
    );
    expect(screen.getByText('foo')).toBeInTheDocument();
    expect(screen.getByText('bar')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it(`should allow mutiple selection on Dropdown when property type is ${PROPERTY_TYPES.ENUM} and the operator is ${OPERATOR_TYPES.IN}`, () => {
    const enumProperty = {
      id: 1,
      name: 'Product Type',
      type: PROPERTY_TYPES.ENUM,
      values: ['foo', 'bar'],
    };
    const operator = OPERATOR_TYPES.IN;
    const props = {
      ...mockProps,
      property: enumProperty,
      operator,
    };
    const { container } = render(<ValueFilter {...props} />);

    expect(screen.getByRole('listbox')).toHaveAttribute(
      'aria-multiselectable',
      'true',
    );
    expect(screen.queryAllByRole('option')).toHaveLength(
      enumProperty.values.length,
    );
    expect(container).toMatchSnapshot();
  });

  it(`should not render options when property type is ${PROPERTY_TYPES.ENUM} doesn't have values`, () => {
    const enumProperty = {
      id: 1,
      name: 'Product Type',
      type: PROPERTY_TYPES.ENUM,
    };
    const operator = OPERATOR_TYPES.IN;
    const props = {
      ...mockProps,
      property: enumProperty,
      operator,
    };
    const { container } = render(<ValueFilter {...props} />);

    expect(screen.getByRole('listbox')).toHaveAttribute(
      'aria-multiselectable',
      'true',
    );
    expect(screen.queryAllByRole('option')).toHaveLength(0);
    expect(container).toMatchSnapshot();
  });

  it(`should render an Input when property type is ${PROPERTY_TYPES.STRING}`, () => {
    const enumProperty = {
      id: 1,
      name: 'Product Name',
      type: PROPERTY_TYPES.STRING,
    };
    const operator = OPERATOR_TYPES.EQUALS;
    const props = {
      ...mockProps,
      property: enumProperty,
      operator,
    };
    const { container } = render(<ValueFilter {...props} />);

    expect(screen.getByRole('textbox')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
