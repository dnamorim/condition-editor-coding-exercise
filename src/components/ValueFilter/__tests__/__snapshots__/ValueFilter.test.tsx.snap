// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueFilter should allow mutiple selection on Dropdown when property type is enumerated and the operator is in 1`] = `
<div>
  <div
    aria-expanded="false"
    aria-multiselectable="true"
    class="ui fluid multiple selection dropdown"
    role="listbox"
    tabindex="0"
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider default text"
      role="alert"
    >
      Value(s)
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition"
    >
      <div
        aria-checked="false"
        aria-selected="true"
        class="selected item"
        role="option"
        style="pointer-events: all;"
      >
        <span
          class="text"
        >
          foo
        </span>
      </div>
      <div
        aria-checked="false"
        aria-selected="false"
        class="item"
        role="option"
        style="pointer-events: all;"
      >
        <span
          class="text"
        >
          bar
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`ValueFilter should not render options when property type is enumerated doesn't have values 1`] = `
<div>
  <div
    aria-expanded="false"
    aria-multiselectable="true"
    class="ui fluid multiple selection dropdown"
    role="listbox"
    tabindex="0"
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider default text"
      role="alert"
    >
      Value(s)
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition"
    />
  </div>
</div>
`;

exports[`ValueFilter should render a Dropdown when property type is enumerated 1`] = `
<div>
  <div
    aria-expanded="false"
    aria-multiselectable="false"
    class="ui fluid selection dropdown"
    role="listbox"
    tabindex="0"
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider default text"
      role="alert"
    >
      Value(s)
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition"
    >
      <div
        aria-checked="false"
        aria-selected="true"
        class="selected item"
        role="option"
        style="pointer-events: all;"
      >
        <span
          class="text"
        >
          foo
        </span>
      </div>
      <div
        aria-checked="false"
        aria-selected="false"
        class="item"
        role="option"
        style="pointer-events: all;"
      >
        <span
          class="text"
        >
          bar
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`ValueFilter should render an Input when property type is string 1`] = `
<div>
  <div
    class="ui fluid input"
  >
    <input
      placeholder="Value(s)"
      type="text"
      value=""
    />
  </div>
</div>
`;
